src/librbd/operation/FlattenRequest.cc

```diff
@@ -40,6 +40,20 @@ class C_FlattenObject : public C_AsyncObjectThrottle<I> {
       return -ERESTART;
     }
 
+    {
+      RWLock::RLocker snap_lock(image_ctx->snap_lock);
+      if (image_ctx->object_map != nullptr) {
+        bool flags_set;
+        int r = image_ctx->test_flags(RBD_FLAG_OBJECT_MAP_INVALID,
+                                      image_ctx->snap_lock, &flags_set);
+        if (r == 0 && !flags_set &&
+            image_ctx->object_map->object_may_exist(m_object_no)) {
+          // can skip because the object already exists
+          return 1;
+        }
+      }
+    }
```

Comment:
@dillaman Did you mean something like this? May be create and use a new ObjectMap method, e.g. `object_exist`?
And I am not sure about PENDING object state, is it safe to skip in that case?

Reply:
@trociny If the object is pending, it means that someone is attempting to delete it (which really shouldn't happen for an overlapping extent). For safety, I would treat pending as you need to perform the zero-byte write.